### PR TITLE
[SPARK-20941] [SQL] Fix SubqueryExec Reuse

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
Before this PR, Subquery reuse does not work. Below are three issues:
- Subquery reuse does not work.  
- It is sharing the same `SQLConf` (`spark.sql.exchange.reuse`) with the one for Exchange Reuse. 
- No test case covers the rule Subquery reuse. 

This PR is to fix the above three issues. 
- Ignored the physical operator `SubqueryExec` when comparing two plans. 
- Added a dedicated conf `spark.sql.subqueries.reuse` for controlling Subquery Reuse
- Added a test case for verifying the behavior

### How was this patch tested?
N/A

### CODE REVIEW

**Comment by gatorsmile** [comment]
@cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 1]
```diff
@@ -25,6 +25,7 @@ import org.apache.spark.sql.catalyst.errors._
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.expressions.codegen._
+import org.apache.spark.sql.catalyst.plans.QueryPlan
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: unnecessary change?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: Removed.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -700,6 +703,33 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
         row => Seq.fill(16)(Row.merge(row, row))).collect().toSeq)
   }
 
+  test("Verify spark.sql.subquery.reuse") {
+    Seq("true", "false").foreach { reuse =>
+      withSQLConf(SQLConf.SUBQUERY_REUSE_ENABLED.key -> reuse) {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: nit: it's more natural to use `Seq(true, false)` and then do `reuse.toString` here.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -700,6 +703,33 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+        val df = sql(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: it will be good if we also check result of this df

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.2

**Comment by carsonwang** [comment]
@gatorsmile , why was this reverted? Are you going to open another PR to fix it?
